cdc_rsync/server_arch.cc

```diff
@@ -20,76 +20,169 @@
+  // Possible values: AMD64, IA64, ARM64, x86
+  if (absl::StrContains(arch_out, "AMD64")) {
+    return ArchType::kWindows_x86_64;
+  }
+
+  return absl::UnimplementedError(
+      absl::StrFormat(kUnsupportedArchErrorFmt, arch_out));
+}
+
+}  // namespace
 
 // static
-ServerArch::Type ServerArch::Detect(const std::string& destination) {
+ServerArch ServerArch::GuessFromDestination(const std::string& destination) {
   // Path starting with ~ or / -> Linux.
   if (absl::StartsWith(destination, "~") ||
       absl::StartsWith(destination, "/")) {
-    return Type::kLinux;
+    LOG_DEBUG("Guessed server arch type Linux based on ~ or /");
+    return ServerArch(ArchType::kLinux_x86_64, /*is_guess=*/true);
   }
 
   // Path starting with C: etc. -> Windows.
   if (!path::GetDrivePrefix(destination).empty()) {
-    return Type::kWindows;
+    LOG_DEBUG("Guessed server arch type Windows based on drive prefix");
+    return ServerArch(ArchType::kWindows_x86_64, /*is_guess=*/true);
   }
 
   // Path with only / -> Linux.
   if (absl::StrContains(destination, "/") &&
       !absl::StrContains(destination, "\\")) {
-    return Type::kLinux;
+    LOG_DEBUG("Guessed server arch type Linux based on forward slashes");
+    return ServerArch(ArchType::kLinux_x86_64, /*is_guess=*/true);
   }
 
   // Path with only \\ -> Windows.
   if (absl::StrContains(destination, "\\") &&
       !absl::StrContains(destination, "/")) {
-    return Type::kWindows;
+    LOG_DEBUG("Guessed server arch type Windows based on backslashes");
+    return ServerArch(ArchType::kWindows_x86_64, /*is_guess=*/true);
   }
 
   // Default to Linux.
-  return Type::kLinux;
+  LOG_DEBUG("Guessed server arch type Linux as default");
+  return ServerArch(ArchType::kLinux_x86_64, /*is_guess=*/true);
 }
 
 // static
-ServerArch::Type ServerArch::LocalType() {
-#if PLATFORM_WINDOWS
-  return ServerArch::Type::kWindows;
-#elif PLATFORM_LINUX
-  return ServerArch::Type::kLinux;
-#endif
+ServerArch ServerArch::DetectFromLocalDevice() {
+  LOG_DEBUG("Detected local device type %s",
+            GetArchTypeStr(GetLocalArchType()));
+  return ServerArch(GetLocalArchType(), /*is_guess=*/false);
+}
+
+// static
+absl::StatusOr<ServerArch> ServerArch::DetectFromRemoteDevice(
+    RemoteUtil* remote_util) {
+  assert(remote_util);
+
+  // Run uname, assuming it's a Linux machine.
+  std::string uname_out;
+  std::string linux_cmd = "uname -sm";
+  absl::Status status =
+      remote_util->RunWithCapture(linux_cmd, "uname", &uname_out, nullptr);
+  if (status.ok()) {
+    LOG_DEBUG("Uname returned '%s'", uname_out);
+    absl::StatusOr<ArchType> type = GetArchTypeFromUname(uname_out);
+    if (type.ok()) {
+      LOG_DEBUG("Detected server arch type %s from uname",
+                GetArchTypeStr(*type));
+      return ServerArch(*type, /*is_guess=*/false);
+    }
+    status = type.status();
+  }
+  LOG_DEBUG("Failed to detect arch type from uname: %s", status.ToString());
```

Comment:
This log message might be confusing to users of the tool. As I understand it when connecting to a Windows machine this message will always be logged and it looks like a potentially fatal error even though it's expected to fail in that case. I would add something like "Continuing with Windows detection" or something along those lines, to make it clear what happens.

Reply:
Good point, I always somewhat assume that cdc_rsync devs are the only ones looking at debug logs, but that's probably not true.